clients/drcachesim/tracer/output.cpp

```diff
@@ -1008,14 +1037,32 @@ process_and_output_buffer(void *drcontext, bool skip_size_cap)
     if (get_local_mode(data) != mode) {
         if (get_local_mode(data) == BBDUP_MODE_L0_FILTER) {
             NOTIFY(0, "Thread %d: filter mode changed\n", dr_get_thread_id(drcontext));
-            // If a switch occurred, then it is possible that the buffer
-            // contains a mix of filtered and unfiltered records. We discard
-            // the buffer by resetting the buffer pointer.
-            buf_ptr = data->buf_base + header_size;
 
-            // Add a FILTER_ENDPOINT marker to indicate that filtering stops here.
-            buf_ptr +=
-                instru->append_marker(buf_ptr, TRACE_MARKER_TYPE_FILTER_ENDPOINT, 0);
+            // If a switch occurred, then it is possible that the buffer
+            // contains a mix of filtered and unfiltered records. We look for the first
+            // unfiltered record and if such a record is found, we insert the
+            // FILTER_ENDPOINT marker before it.
+            byte *end =
+                (byte *)find_unfiltered_record(data->buf_base + header_size, buf_ptr);
+            if (end == NULL) {
+                // Add a FILTER_ENDPOINT marker to indicate that filtering stops here.
+                buf_ptr +=
+                    instru->append_marker(buf_ptr, TRACE_MARKER_TYPE_FILTER_ENDPOINT, 0);
+            } else {
+                // Write the filtered data.
+                write_trace_data(drcontext, data->buf_base, end, get_local_window(data));
```

Comment:
Here and below should use output_buffer() to support online, to support handoff_buf, and to update the various counts -- was there a reason these can't use output_buffer()?

Reply:
Changed to use output_buffer()